_posts/2018-10-15-12-2-Use-SkipToken-for-Server-Side-Paging.md

```diff
@@ -0,0 +1,63 @@
+---
+layout: post
+title: "12.2 Use $skiptoken for server side paging "
+description: "WebAPI to use $skiptoken for server side paging"
+category: "12. Design"
+---
+# Use $skiptoken for server-driven paging
+
+### Background
+Loading large data can be slow. Services often rely on pagination to load the data incrementally to improve the response times and the user experience. Paging can be server-driven or client-driven:
+#### Client-driven paging
+In client-driven paging, the client decides how many records it wants to load and asks the server that many records. That is achieved by using $skip and $top tokens in conjunction. For instance, if a client needs to request 10 records from 71-80, it can send a similar request as below:
+
+`GET ~/Products/$skip=70&$top=10`
+#### Server-driven paging
+In server-driven paging, the client asks for a collection of entities and the server sends back partial results as well as a nextlink to use to retrieve more results. The nextlink is an opaque link which may use $skiptoken to identify the last loaded record.
+### Problem
+Currently, WebAPI uses $skip for server-driven paging which is a slight deviation from the OData standard and can be problematic when the data source can get updated concurrently. For instance, a deletion of a record may cause the last record to be sent down to the client twice. 
+### Proposed Solution
+WebAPI will now implement $skiptoken. When a collection of entity is requested which requires paging, we will assign the key value of the last sent entity to $skiptoken in the nextlink url. While processing a request with $skiptoken, we will add another condition (the key of the entity to be greater than the value specified to the skiptoken) to the predicate. 
```

Comment:
Adding this condition for $skiptoken processing: how to accommodate the case for $orderby with non-key attribute, i.e, the query result returned are not ordered by key attributes?

Reply:
All the other query options will be preserved except for $skip.

Reply:
Design updated.